Added Download Dropdown to Series LTI-Tools

Arnei opened the pull request:
This PR allows users to pass a new URL parameter "download". It enables a download button in the Series view on each episode.
The button will open a dropdown that allows the user to download individual tracks from that episode.

The PR is sponsored by the University of Ulm

![Bildschirmfoto vom 2021-03-22 14-39-47](https://user-images.githubusercontent.com/14070005/111998638-831d1b00-8b1c-11eb-919b-4a12b7ad2b0f.png)


### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [ ] be against the correct branch (features can only go into develop)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

Arnei committed:
Added Download Dropdown to Series LTI-Tools

Users can now pass an URL parameter to enable a download button in the Series view on each episode.
The button will open a dropdown that allows the user to download individual tracks from that episode.

wsmirnow commented:
I've got an error `Fehler: Cannot read property 'resolution' of undefined`.  The episodes contains adaptive streaming publications. These tracks looks like

```
{
  "id": "track-1",
  "type": "presentation/delivery",
  "ref": "track:reference-id",
  "transport": "HLS",
  "mimetype": "application/x-mpegURL",
  "tags": {
    "tag": ["360p-quality", "engage-download", "engage-streaming"]
  },
  "url": "https://streaming.server.edu/opencast/smil:engage-player_track-1_presentation.smil/playlist.m3u8",
  "duration": 900052,
  "live": false
}
```
The video tag is missing.

Arnei committed:
Added handling for adaptive streaming publications

- Tries to exclude tracks from adapative streaming publications from the search results
- Error handling for tracks without a resolution

Arnei commented:
> I've got an error `Fehler: Cannot read property 'resolution' of undefined`. The episodes contains adaptive streaming publications. 

Thanks for the quick feedback. I have added error handling for tracks that are missing the video tag.

As per suggestion I have also tried to exclude tracks from adaptive streaming publications in the list of downloadable tracks.

Arnei committed:
Fixed check for resolution again

Now only adds tracks to the download menu that contain a video element with a resolution. This is also used to avoid adaptive streaming tracks.

Arnei committed:
Removed the the role "Instructor" requirement for download button

Arnei commented:
Removed the check for the LTI role "Instructor", so that every LTI authenticated person can see the button. Reasoning is that the users of this button are most likely not gonna be the instructors, but the students, who typically rank lowest on the ranking list of access rights.

pascalseeland commented on modules/lti/src/utils.ts:
```diff
@@ -1,5 +1,75 @@
+ * If tracks have the same type, they are then sorted by resolution
+ * @param tracks the tracks to be sorted
+ */
+export const sortByType = (tracks: Track[]) => {
```
I would prefer to move this function into its own ts file. Something like trackUtils.ts. As such, utils will only contain general usable functions and track specific stuff can go into trackUtils

wsmirnow replied:
I am pretty sure @KatrinIhler is not a fan of utility classes, that never been used as intended. I share her opinion. If we have a specific case, where we can reuse this function, then we can move it out.

wsmirnow replied:
Ok, It is currently in a utils file, sorry for noise. But do you really want more utils files? really? :D

wsmirnow replied:
Ok, It is currently in a utils file, sorry for noise. But do you really want more utils files? really? :D

KatrinIhler replied:
I have been summoned? :D (It's especially funny since I was just talking to Malte about his utils file. :rofl:)

I can't be bothered to look at this closely right now, so just a general request: no more utility files than absolutely necessary, and if you can name them better than something-utils, please do. :D

pascalseeland replied:
The main reason for my request was because I did not want to have a utils file which contains a lot of miscellaneous function.

pascalseeland commented on modules/lti/src/utils.ts:
```diff
@@ -1,5 +1,75 @@
+ * @param descA resolution of the first track
+ * @param descB resolution of the second track
+ */
+const sortResolutions = (descA: string, descB: string) => {
```
Either also move this function to trackUtils or make it a general function for comparing two resolutions. Also I would prefer to rename it compareResolutions, as it does not sort them, like the sortTracks above

pascalseeland reviewed (changes requested):
Please take a look at utils.ts. Otherwise this is fine

wsmirnow reviewed (changes requested):
Can you document the new `download` option in `/docs/guides/admin/docs/modules/ltimodule.md` please.

Arnei committed:
Cleaned up util functions

- Created a more specific utils file for tracks
- Rewrote resolution comparison to be more generic
- Introduced some type safety for resolutions

Arnei committed:
Added lti-series download button info to admin docs

KatrinIhler commented:
Approved twice, so I'm gonna merge this.
